### ISSUE TITLE
[FR]: Date format wishes

### ISSUE THREAD

**Issue by Owyn** [martinrotter/rssguard#608]
### Brief description of the feature request

1) ability to show date in **relative time** instead of exact one

e.g. instead of 
`21-01-2022 16:00`
we could have
`30 minutes ago`
`4 hours ago`
`1 day ago`

~~**2) Ability to hide the year would be very welcomed (I don't wake up each day forgetting which year it is today so no need to write it each n every time...)**~~

e.g.
`21-01 16:00`
instead of
`21-01-2022 16:00`

~~**3) ability to hide exact time for items which are older than today**~~

e.g. instead of 
`21-01-2022 16:00`
just show
`21-01-2022`

**Comment by martinrotter** [comment]
2. is already there, see settings.
![image](https://user-images.githubusercontent.com/1255302/152653945-553fda73-8f26-49b9-89aa-cddbe52c346f.png)

**Comment by Owyn** [comment]
but on your screenshot the year it still shown?...

same with settings, there are no variants with no year:
![image](https://user-images.githubusercontent.com/1309656/152654077-7a8f06f2-fb13-447e-adaa-77776d26f63f.png)

**Comment by martinrotter** [comment]
on screenshot today articles only have time, but I agree, setting sucks and could indeed not offer time-only variants, that is a bug will fix

**Comment by Owyn** [comment]
I don't mean todays date format, I mean date format for everything BUT today to not have year in it

**Comment by martinrotter** [comment]
I see, OK. I made those formats of date/time editable, you can now fully customize, formatted date automatically shows in tooltip.

![image](https://user-images.githubusercontent.com/1255302/152670442-7cc74315-9c11-4eb0-b60f-e1f366f61e76.png)

![image](https://user-images.githubusercontent.com/1255302/152670459-1068a48b-7c30-4a8a-baaf-31c441a31b0d.png)


https://doc.qt.io/qt-5/qtime.html#toString
https://doc.qt.io/qt-5/qdate.html#toString-2

**Comment by martinrotter** [comment]
8cb7c1c

**Comment by Owyn** [comment]
There are tooltips when format is already selected but there are no tooltips in the selection list (so currently to see results of all formats - you'd have to choose each of them - tons of clicking)
![image](https://user-images.githubusercontent.com/1309656/152675164-7c1ae634-fa3e-4b30-b76b-eecb214ee5db.png)

**Comment by martinrotter** [comment]
4616488 - added tooltips for all items

**Comment by martinrotter** [comment]
added more time formats

**Comment by martinrotter** [comment]
> 1. ability to show date in **relative time** instead of exact one

Do you want to have format of this configurable?

**Comment by Owyn** [comment]
> > 1. ability to show date in **relative time** instead of exact one
> 
> Do you want to have format of this configurable?

Even "300 days ago, 3 years ago" would be more informative than just a plain date 01.02.03

but an option
`"Show relative time for items not older than [ ] days"`
would be nice to display relative time only for items of this week (last 7 days) for example

**Comment by martinrotter** [comment]
So, it should show "relative" time for articles NEWER than limit?

**Comment by Owyn** [comment]
> So, it should show "relative" time for articles NEWER than limit?

Yes, only of current week (last 7 days) for example

**Comment by martinrotter** [comment]
Soé something like this?

![image](https://user-images.githubusercontent.com/1255302/153877587-840aafcd-9c5f-422c-a2ee-9233be983a02.png)

**Comment by martinrotter** [comment]
![image](https://user-images.githubusercontent.com/1255302/153877815-d708bfdd-a0ff-4929-a419-39475f84fead.png)

**Comment by Owyn** [comment]
Yes, exactly like that.

Like most sites write time for the stuff they post:
![image](https://user-images.githubusercontent.com/1309656/153878649-dcda3c33-0e20-4bb1-88e6-4097fd0b2e49.png)

**Comment by martinrotter** [comment]
Is the format OK? "x minutes/hours/days/weeks ago"?

**Comment by martinrotter** [comment]
OK, i pushed it, it can be tweaked in settings and is disabled by default. will polish the feature tomorrown.

**Comment by Owyn** [comment]
You stopped at `x weeks ago` I see 😄 

I hope we'd get `months ago` as well (many of my items come out monthly)

When is the time updated btw?

**Comment by Owyn** [comment]
also, an option to stop at `days` format would be nice I think
e.g.
`8 days ago`
`35 days ago`

**Comment by AmmarSaleemG** [comment]
So, if you are showing relative time, how do you deal with this:

![image](https://user-images.githubusercontent.com/15247707/153924408-f2009b9d-dacf-4ed2-8a8b-4f2568305efc.png)

Sometimes, dates of articles from "The Economist" and "Geo News(local news channel)" are shown ahead of the current time (Probably my UTC is +5, and it is not aligning with them).

Is there anything that can force (or subtract) the time of an article posted ahead of the time?

Posting link of feeds that have conflict of timing:
http://feeds.feedburner.com/GeoNewsOfficial
https://www.economist.com/international/rss.xml

**Comment by Owyn** [comment]
> ahead of the current time

low quality feeds from the future 🤷🏻 

if they are shown with the time 50/120/285 minutes into the future, e.g. now is 12:00 and their time is 12:50+ - how do you deal with it now?

maybe just count all that as an "incorrect date format" and set current time instead?

**Comment by martinrotter** [comment]
> So, if you are showing relative time, how do you deal with this:
> 
> ![image](https://user-images.githubusercontent.com/15247707/153924408-f2009b9d-dacf-4ed2-8a8b-4f2568305efc.png)
> 
> Sometimes, dates of articles from "The Economist" and "Geo News(local news channel)" are shown ahead of the current time (Probably my UTC is +5, and it is not aligning with them).
> 
> Is there anything that can force (or subtract) the time of an article (to the current timestamp) posted ahead of the time?
> 
> Posting link of feeds that have conflict of timing: http://feeds.feedburner.com/GeoNewsOfficial https://www.economist.com/international/rss.xml

ROFL, that literally made my day. I actually inticipated that potential regression. :D Hahaha. Sorry.

>low quality feeds from the future 🤷🏻

Haha, sorry. My belly hurts from the hardest laugh if this year I have had so far. Jesus Christ.

EDIT: I will look into it soon.

**Comment by martinrotter** [comment]
Just checked your first faulty feed link.

One article has time `Posted:Mon, 14 Feb 2022 22:58:00 +0000` (UTC) while current UTC time is `18:50:37 UTC Monday, 14 February 2022`.

OK. This is probably THE most shitty feed I have ever seen and I must say that I am atheist, but holy shit.

**Comment by AmmarSaleemG** [comment]
> I have ever seen and I must say that I am atheist, but holy shit.

LMFAO, wait till you get a feed from The Economist. Which is two days in the future.

**Comment by martinrotter** [comment]
Actually, I do not see any problems in economist one.

**Comment by AmmarSaleemG** [comment]
> Actually, I do not see any problems in economist one.
> 
> Newest article from that feed has: `<pubDate>Sat, 12 Feb 2022 00:00:00 +0000</pubDate>`

Yeah, this feed was posted on 10th Feb and stayed on top till 12th Feb.
I was asking to wait few days till they publish the next feed.

**Comment by AmmarSaleemG** [comment]
TYPO:
-----------------------
![image](https://user-images.githubusercontent.com/15247707/154068309-f19c44a3-8089-4a4a-95af-96d9f54f7ec0.png)

1 Hour instead of 1 Hour**s**
or add ```n Hour(s) ago```

**Comment by Owyn** [comment]
I can't enter more than 2 digit into the setting of days, only use scroller
![image](https://user-images.githubusercontent.com/1309656/154456818-b400131f-f654-4b7c-b62d-2389ce2b12d4.png)

**Comment by martinrotter** [comment]
> ## TYPO:
> 
> ![image](https://user-images.githubusercontent.com/15247707/154068309-f19c44a3-8089-4a4a-95af-96d9f54f7ec0.png)
> 
> 1 Hour instead of 1 Hour**s** or add `n Hour(s) ago`
> 
> EDIT: Also add an option for abbreviating them too, e.g.:
> 
>     * Minute(s) to Min(s)
> 
>     * Hour(s) to Hr(s)
> 
>     * Second(s) to Sec(s)

no gonna happen sry. even current state can be vastly configured when compared to some other readers

**Comment by martinrotter** [comment]
> I can't enter more than 2 digit into the setting of days, only use scroller ![image](https://user-images.githubusercontent.com/1309656/154456818-b400131f-f654-4b7c-b62d-2389ce2b12d4.png)

cannot reproduce, sry, i can just click in and type in any number, note that 300 is max

**Comment by Owyn** [comment]
> 300 is max

why the limit tho?

ah - that might explain it, - cuz try entering "999" or "444" - only first two digits would show up

**Comment by Owyn** [comment]
maybe make day limit an option tho:

Options:
1) Show relative time [x]
1.1) Only for articles not older than [x] days
1.2) Always show format in `days ago` (even for 7+ days old items)

**Comment by Owyn** [comment]
![image](https://user-images.githubusercontent.com/1309656/154806790-2ae93808-93dc-4ac1-b3e4-040824b73d45.png)

Maybe write "Just now" or smth like that instead of "0 minutes ago" ?

**Comment by martinrotter** [comment]
> ![image](https://user-images.githubusercontent.com/1309656/154806790-2ae93808-93dc-4ac1-b3e4-040824b73d45.png)
> 
> Maybe write "Just now" or smth like that instead of "0 minutes ago" ?

added, everything newer than 120 seconds from now will have "just now"

**Comment by martinrotter** [comment]
Also switched to priority of those two formatting options "todays format" and "relative format", so that now they actually can work both at the same time.

![image](https://user-images.githubusercontent.com/1255302/154922196-0d14eb7d-3ddb-4f57-be6f-8c34c1ef0599.png)

**Comment by Owyn** [comment]
![image](https://user-images.githubusercontent.com/1309656/154934794-df7a11fb-10da-4c63-a719-b450149dd7aa.png)


![image](https://user-images.githubusercontent.com/1309656/154934724-f6aa275c-297c-49f4-97a8-4f21addd8dbe.png)

**Comment by martinrotter** [comment]
change was not pushed

14f8f3bcc

**Comment by AmmarSaleemG** [comment]
![image](https://user-images.githubusercontent.com/15247707/155667835-5380fc99-f6ea-4d47-9b97-e52a6052dcf7.png)


### The "Just Now" rhetoric _(for low quality feeds)_ looks lame.
### Better if it has "In x (Minutes | Hours | Days)"

**Comment by martinrotter** [comment]
> ![image](https://user-images.githubusercontent.com/15247707/155667835-5380fc99-f6ea-4d47-9b97-e52a6052dcf7.png)
> 
> Today's date is 25 and this one is set to publish on 26th.
> #### The "Just Now" rhetoric _(for low quality feeds)_ looks lame. Better if it has "In x (Minutes | Hours | Days)"

No, I will not adjust RSS Guard behavior to work with shit quality feeds this way, sorry.

**Comment by Owyn** [comment]
Why not just treat those dates properly as "invalid date" and set current date for them?

**Comment by AmmarSaleemG** [comment]
> No, I will not adjust RSS Guard behavior to work with shit quality feeds this way, sorry.

@martinrotter Actually, i would have never said it, but it is annoying to read "Just Now" perpetually. 
Setting relative time requires such a workaround that it is imperative to make it flexible back and forth. 
Our focus is mainly on content and not time But yet, this feed is always put at the top of the list, which annoys you more.

I have a little suggestion if you allow me to give: While fetching the feed, Read the timestamp, if it is ahead of time, substitute it with the current timestamp, and that's it, will be more relevant. 



> Why not just treat those dates properly as "invalid date" and set current date for them?
> 
> cuz if "just now" stays there for days it's kind of a bug (it should stay there for a minute-two)

@Owyn "Invalid Date" will also be unprofessional for such a flexible and multifunctional feed reader.

**Comment by martinrotter** [comment]
> I have a little suggestion if you allow me to give: While fetching the feed, Read the timestamp, if it is ahead of time, substitute it with the current timestamp, and that's it, will be more relevant.

You can easily create article filter to do that. No need to add extra code into RSS Guard.

```js
function filterMessage() {
  if (msg.created > new Date()) {
    msg.created = new Date();
  }

  return MessageObject.Accept;
}
```

**Comment by Owyn** [comment]
> No need to add extra code into RSS Guard.

It's still an `invalid date` there in the feed tho

**Comment by Owyn** [comment]
just like when the date is missing or it's simply written `"hello"` there in the `date` field - it's invalid, RSSGuard probably shouldn't let invalid data through

**Comment by martinrotter** [comment]
> just like when the date is missing or it's simply written `"hello"` there in the `date` field - it's invalid, RSSGuard probably shouldn't let invalid data through - because it might break the functionality of the app - as shown above for example "just now" would stay there for more than intended and break sorting

1. I fixed your request: d4e4abfa8 + 5c1b5907e.
2. There is a LONG (trust me, LONG) history of dealing with feed entries with invalid data, and "date" field is just one of those. RSS Guard behaves a bit differently if date/time was not actually fetched from feed (because that info is not there or has wrong format etc.), but replaced with current time.

> "just now" would stay there for more than intended and break sorting

No, it would not, fixed with 1. Feature is switchable in settings -> articles.

**Comment by AmmarSaleemG** [comment]
> You can easily create article filter to do that. No need to add extra code into RSS Guard.
```javascript
function filterMessage() {
  if (msg.created > new Date()) {
    msg.created = new Date();
  }

  return MessageObject.Accept;
}
```
This code kept refreshing the feed time until the original time of feed. Hence, it broke the sorting by staying on top.

> * I fixed your request: [d4e4abf](https://github.com/martinrotter/rssguard/commit/d4e4abfa8441173b668330a8ca9a98b26b369b50) + [5c1b590](https://github.com/martinrotter/rssguard/commit/5c1b5907ee75c24584deb0948e4125ea3bfc93b0).

This is working as we intended.

**Comment by martinrotter** [comment]
> > You can easily create article filter to do that. No need to add extra code into RSS Guard.
> 
> ```js
> function filterMessage() {
>   if (msg.created > new Date()) {
>     msg.created = new Date();
>   }
> 
>   return MessageObject.Accept;
> }
> ```
> 
> This code kept refreshing the feed time until the original time of feed. Hence, it broke the sorting by staying on top.
> 
> > * I fixed your request: [d4e4abf](https://github.com/martinrotter/rssguard/commit/d4e4abfa8441173b668330a8ca9a98b26b369b50) + [5c1b590](https://github.com/martinrotter/rssguard/commit/5c1b5907ee75c24584deb0948e4125ea3bfc93b0).
> 
> This is working as we intended.

Yes, the filter solution would need one extra line: 

`msg.createdIsMadeup = true;`
